File: js/src/internal/aes_siv.ts
// Copyright (C) 2017 Dmitry Chestnykh, Tony Arcieri
// MIT License. See LICENSE file for details.

import { equal } from "./constant-time";
import { wipe } from "./wipe";
import { xor } from "./xor";

import IntegrityError from "../exceptions/integrity_error";
import NotImplementedError from "../exceptions/not_implemented_error";
import Block from "./block";
import { ICtrLike, IMacLike, ISivLike } from "./interfaces";

import PolyfillCrypto from "./polyfill";
import PolyfillAes from "./polyfill/aes";
import PolyfillAesCmac from "./polyfill/aes_cmac";
import PolyfillAesCtr from "./polyfill/aes_ctr";
import WebCryptoAesCmac from "./webcrypto/aes_cmac";
import WebCryptoAesCtr from "./webcrypto/aes_ctr";

/** Maximum number of associated data items */
const MAX_ASSOCIATED_DATA = 126;

/** The AES-SIV mode of authenticated encryption */
export default class AesSiv implements ISivLike {
  /** Create a new AesSiv instance with the given 32-byte or 64-byte key */
  public static async importKey(
    keyData: Uint8Array,
    crypto: Crypto | PolyfillCrypto,
  ): Promise<AesSiv> {
    // We only support AES-128 and AES-256. AES-SIV needs a key 2X as long the intended security level
    if (keyData.length !== 32 && keyData.length !== 64) {
      throw new Error(`AES-SIV: key must be 32 or 64-bits (got ${keyData.length}`);
    }

    const macKey = keyData.subarray(0, keyData.length / 2 | 0);
    const encKey = keyData.subarray(keyData.length / 2 | 0);

    if (crypto instanceof PolyfillCrypto) {
      const mac = new PolyfillAesCmac(new PolyfillAes(macKey));
      const ctr = new PolyfillAesCtr(new PolyfillAes(encKey));
      return new AesSiv(mac, ctr, null);
    } else {
      const mac = await WebCryptoAesCmac.importKey(macKey, crypto);

      try {
        const ctr = await WebCryptoAesCtr.importKey(encKey, crypto);
        return new AesSiv(mac, ctr, crypto);
      } catch (e) {
        if (e.message.includes("unsupported")) {
          throw new NotImplementedError("AES-SIV: unsupported crypto backend (CTR missing). Use PolyfillCrypto.");
        } else {
          throw e;
        }
      }
    }
  }

  private _mac: IMacLike;
  private _ctr: ICtrLike;
  private _tmp1: Block;
  private _tmp2: Block;
  private _crypto: Crypto | null;

  constructor(mac: IMacLike, ctr: ICtrLike, crypto: Crypto | null) {
    this._mac = mac;
    this._ctr = ctr;
    this._crypto = crypto;
    this._tmp1 = new Block();
    this._tmp2 = new Block();
  }

  /** Encrypt and authenticate data using AES-SIV */
  public async seal(plaintext: Uint8Array, associatedData: Uint8Array[]): Promise<Uint8Array> {
    if (associatedData.length > MAX_ASSOCIATED_DATA) {
      throw new Error("AES-SIV: too many associated data items");
    }

    // Allocate space for sealed ciphertext.
    const resultLength = Block.SIZE + plaintext.length;
    const result = new Uint8Array(resultLength);

    // Authenticate.
    const iv = await this._s2v(associatedData, plaintext);
    result.set(iv);

    // Encrypt.
    zeroIVBits(iv);
    result.set(await this._ctr.encrypt(iv, plaintext), iv.length);
    return result;
  }

  /** Decrypt and authenticate data using AES-SIV */
  public async open(sealed: Uint8Array, associatedData: Uint8Array[]): Promise<Uint8Array> {
    if (associatedData.length > MAX_ASSOCIATED_DATA) {
      throw new Error("AES-SIV: too many associated data items");
    }

    if (sealed.length < Block.SIZE) {
      throw new IntegrityError("AES-SIV: ciphertext is truncated");
    }

    // Decrypt.
    const tag = sealed.subarray(0, Block.SIZE);
    const iv = this._tmp1.data;
    iv.set(tag);
    zeroIVBits(iv);

    const result = await this._ctr.decrypt(iv, sealed.subarray(Block.SIZE));

    // Authenticate.
    const expectedTag = await this._s2v(associatedData, result);

    if (!equal(expectedTag, tag)) {
      wipe(result);
      throw new IntegrityError("AES-SIV: ciphertext verification failure!");
    }

    return result;
  }

  /** Make a best effort to wipe memory used by this AesSiv instance */
  public clear(): this {
    this._tmp1.clear();
    this._tmp2.clear();
    this._ctr.clear();
    this._mac.clear();

    return this;
  }

  /**
   * The S2V operation consists of the doubling and XORing of the outputs
   * of the pseudo-random function CMAC.
   *
   * See Section 2.4 of RFC 5297 for more information
   */
  private async _s2v(associated_data: Uint8Array[], plaintext: Uint8Array): Promise<Uint8Array> {
    this._mac.reset();
    this._tmp1.clear();

    // Note: the standalone S2V returns CMAC(1) if the number of passed
    // vectors is zero, however in SIV construction this case is never
    // triggered, since we always pass plaintext as the last vector (even
    // if it's zero-length), so we omit this case.
    await this._mac.update(this._tmp1.data);
    this._tmp2.clear();
    this._tmp2.data.set(await this._mac.finish());
    this._mac.reset();

    for (const ad of associated_data) {
      await this._mac.update(ad);
      this._tmp1.clear();
      this._tmp1.data.set(await this._mac.finish());
      this._mac.reset();
      this._tmp2.dbl();
      xor(this._tmp2.data, this._tmp1.data);
    }

    this._tmp1.clear();

    if (plaintext.length >= Block.SIZE) {
      const n = plaintext.length - Block.SIZE;
      this._tmp1.data.set(plaintext.subarray(n));
      await this._mac.update(plaintext.subarray(0, n));
    } else {
      this._tmp1.data.set(plaintext);
      this._tmp1.data[plaintext.length] = 0x80;
      this._tmp2.dbl();
    }
    xor(this._tmp1.data, this._tmp2.data);
    await this._mac.update(this._tmp1.data);
    return this._mac.finish();
  }
}

/** Zero out the top bits in the last 32-bit words of the IV */
function zeroIVBits(iv: Uint8Array) {
  // "We zero-out the top bit in each of the last two 32-bit words
  // of the IV before assigning it to Ctr"
  //  — http://web.cs.ucdavis.edu/~rogaway/papers/siv.pdf
  iv[iv.length - 8] &= 0x7f;
  iv[iv.length - 4] &= 0x7f;
}

File: dotnet/Miscreant/AesSiv.cs
using System;
using System.Security.Cryptography;

namespace Miscreant
{
	/// <summary>
	/// AES-SIV authenticated encryption mode, defined in
	/// <see href="https://tools.ietf.org/html/rfc5297">RFC 5297</see>.
	/// </summary>
	public sealed class AesSiv : IDisposable
	{
		private const int BlockSize = Constants.BlockSize;
		private const int MaxAssociatedDataItems = 126;

		private static readonly byte[] Empty = new byte[0];
		private static readonly byte[] Zero = new byte[BlockSize];

		private readonly IMac mac;
		private readonly AesCtr ctr;
		private bool disposed;

		private AesSiv(Func<byte[], IMac> macFactory, byte[] key)
		{
			if (key == null)
			{
				throw new ArgumentNullException(nameof(key));
			}

			if (key.Length != Constants.AesSiv256KeySize && key.Length != Constants.AesSiv512KeySize)
			{
				throw new CryptographicException("Specified key is not a valid size for this algorithm.");
			}

			int halfKeySize = key.Length / 2;

			var K1 = new byte[halfKeySize];
			var K2 = new byte[halfKeySize];

			Array.Copy(key, 0, K1, 0, halfKeySize);
			Array.Copy(key, halfKeySize, K2, 0, halfKeySize);

			mac = macFactory(K1);
			ctr = new AesCtr(K2);
		}

		/// <summary>
		/// Initializes a new instance of the AES-CMAC-SIV algorithm with the specified key.
		/// </summary>
		/// <param name="key">The secret key for AES-CMAC-SIV encryption.</param>
		/// <returns>An AES-CMAC-SIV instance.</returns>
		public static AesSiv CreateAesCmacSiv(byte[] key)
		{
			return new AesSiv(AesCmac.Create, key);
		}

		/// <summary>
		/// Initializes a new instance of the AES-PMAC-SIV algorithm with the specified key.
		/// </summary>
		/// <param name="key">The secret key for AES-PMAC-SIV encryption.</param>
		/// <returns>An AES-PMAC-SIV instance.</returns>
		public static AesSiv CreateAesPmacSiv(byte[] key)
		{
			return new AesSiv(AesPmac.Create, key);
		}

		/// <summary>
		/// Seal encrypts and authenticates plaintext, authenticates the given
		/// associated data items, and returns the result. For nonce-based
		/// encryption, the nonce should be the last associated data item.
		/// </summary>
		/// <param name="plaintext">The plaintext to encrypt.</param>
		/// <param name="data">Associated data items to authenticate.</param>
		/// <returns>Concatenation of the authentication tag and the encrypted data.</returns>
		public byte[] Seal(byte[] plaintext, params byte[][] data)
		{
			if (data == null)
			{
				throw new ArgumentNullException(nameof(data));
			}

			if (data.Length > MaxAssociatedDataItems)
			{
				throw new CryptographicException($"Maximum number of associated data items is {MaxAssociatedDataItems}");
			}

			if (plaintext == null)
			{
				plaintext = Empty;
			}

			byte[] iv = S2V(data, plaintext);
			byte[] output = new byte[iv.Length + plaintext.Length];

			Array.Copy(iv, output, iv.Length);
			ZeroIvBits(iv);

			ctr.Reset(iv);
			ctr.Encrypt(plaintext, 0, plaintext.Length, output, iv.Length);

			return output;
		}

		/// <summary>
		/// Open decrypts ciphertext, authenticates the decrypted plaintext
		/// and the given associated data items and, if successful, returns
		/// the result. For nonce-based encryption, the nonce should be the
		/// last associated data item. In case of failed decryption, this
		/// method throws <see cref="CryptographicException">.
		/// </summary>
		/// <param name="ciphertext">The ciphertext to decrypt.</param>
		/// <param name="data">Associated data items to authenticate.</param>
		/// <returns>The decrypted plaintext.</returns>
		/// <exception cref="CryptographicException">Thrown when the ciphertext is invalid.</exception>
		public byte[] Open(byte[] ciphertext, params byte[][] data)
		{
			if (ciphertext == null)
			{
				throw new ArgumentNullException(nameof(ciphertext));
			}

			if (data == null)
			{
				throw new ArgumentNullException(nameof(data));
			}

			if (ciphertext.Length < BlockSize)
			{
				throw new CryptographicException("Malformed or corrupt ciphertext.");
			}

			if (data.Length > MaxAssociatedDataItems)
			{
				throw new CryptographicException($"Maximum number of associated data items is {MaxAssociatedDataItems}");
			}

			byte[] iv = new byte[BlockSize];
			byte[] output = new byte[ciphertext.Length - iv.Length];

			Array.Copy(ciphertext, 0, iv, 0, BlockSize);
			ZeroIvBits(iv);

			ctr.Reset(iv);
			ctr.Encrypt(ciphertext, BlockSize, output.Length, output, 0);

			byte[] v = S2V(data, output);

			if (!Subtle.ConstantTimeEquals(ciphertext, v, BlockSize))
			{
				throw new CryptographicException("Malformed or corrupt ciphertext.");
			}

			return output;
		}

		/// <summary>
		/// S2V operation, defined in the section 2.4 of
		/// <see href="https://tools.ietf.org/html/rfc5297#section-2.4">RFC 5297</see>.
		/// </summary>
		private byte[] S2V(byte[][] headers, byte[] message)
		{
			if (headers == null)
			{
				throw new ArgumentNullException(nameof(headers));
			}

			if (message == null)
			{
				throw new ArgumentNullException(nameof(message));
			}

			// The standalone S2V returns CMAC(1) if the number of
			// passed vectors is zero, however in SIV contruction
			// this case is never triggered, since we always pass
			// plaintext as the last vector (even if it's zero-length),
			// so we omit this case.

			mac.HashCore(Zero, 0, BlockSize);
			byte[] v = mac.HashFinal();

			foreach (var header in headers)
			{
				if (header == null)
				{
					continue;
				}

				mac.HashCore(header, 0, header.Length);
				Utils.Multiply(v);
				Utils.Xor(mac.HashFinal(), v, BlockSize);
			}

			if (message.Length > BlockSize)
			{
				int n = message.Length - BlockSize;

				mac.HashCore(message, 0, n);
				Utils.Xor(message, n, v, 0, BlockSize);
				mac.HashCore(v, 0, BlockSize);

				return mac.HashFinal();
			}

			byte[] padded = new byte[BlockSize];

			Array.Copy(message, padded, message.Length);
			Utils.Multiply(v);
			Utils.Pad(padded, message.Length);
			Utils.Xor(padded, v, BlockSize);
			mac.HashCore(v, 0, BlockSize);

			return mac.HashFinal();
		}

		private void ZeroIvBits(byte[] iv)
		{
			iv[iv.Length - 8] &= 0x7f;
			iv[iv.Length - 4] &= 0x7f;
		}

		public void Dispose()
		{
			if (!disposed)
			{
				mac.Dispose();
				ctr.Dispose();

				disposed = true;
			}
		}
	}
}
